exchanges/exchanges.go

```diff
@@ -3133,11 +3221,139 @@ func (dcr *DecredDEX) processWsMessage(raw []byte) {
+						cache.lastStamp = candle.EndStamp
+						cache.mtx.Unlock()
+					} else {
+						// Our candles are out of sync. Get a fresh set.
+						cacheKey := cache.key
+						cache.mtx.Unlock()
+						dcr.clearCandleCache(binSize)
+						dcr.request(msgjson.CandlesRoute, &msgjson.CandlesRequest{
+							BaseID:     42,
+							QuoteID:    0,
+							BinSize:    (time.Duration(binSize) * time.Millisecond).String(),
+							NumCandles: dexapi.CacheSize,
+						}, func(msg *msgjson.Message) {
+							dcr.handleCandles(cacheKey, msg)
+						})
+					}
+				}
+			}
 		}
 	}
 	dcr.wsUpdated()
 }
 
+// convertDEXCandle converts the *msgjson.Candle to a dexdb.Candle.
+func convertDEXCandle(c *msgjson.Candle) *dexdb.Candle {
+	return &dexdb.Candle{
+		StartStamp:  c.StartStamp,
+		EndStamp:    c.EndStamp,
+		MatchVolume: c.MatchVolume,
+		QuoteVolume: c.QuoteVolume,
+		HighRate:    c.HighRate,
+		LowRate:     c.LowRate,
+		StartRate:   c.StartRate,
+		EndRate:     c.EndRate,
+	}
+}
+
+// handleSubResponse handles the response to the order book subscription.
+func (dcr *DecredDEX) handleSubResponse(msg *msgjson.Message) {
+	ob := new(msgjson.OrderBook)
+	err := msg.UnmarshalResult(ob)
+	if err != nil {
+		dcr.setWsFail(fmt.Errorf("error unmarshaling orderbook response: %v", err))
+	}
+	dcr.setOrderBook(ob)
+}
+
+// handleCandles handles the response for a set of candles from the data API.
+func (dcr *DecredDEX) handleCandles(key candlestickKey, msg *msgjson.Message) {
+	wireCandles := new(msgjson.WireCandles)
+	err := msg.UnmarshalResult(wireCandles)
+	if err != nil {
+		log.Errorf("Error unmarshaling candlestick response from DEX at %s: %v", dcr.cfg.Host, err)
+		return
+	}
+
+	binSize := uint64(key.duration().Milliseconds())
+
+	candles := wireCandles.Candles()
+
+	cache := &candleCache{
+		CandleCache: dexdb.NewCandleCache(len(candles), binSize),
+		key:         key,
+	}
+
+	for _, candle := range candles {
+		cache.Add(convertDEXCandle(candle))
+	}
+	if len(candles) > 0 {
+		cache.lastStamp = candles[len(candles)-1].EndStamp
+	}
+	dcr.setCandleCache(binSize, cache)
+}
+
+// handleConfigResponse handles the response for the DEX configuration.
+func (dcr *DecredDEX) handleConfigResponse(msg *msgjson.Message) {
+	cfg := new(msgjson.ConfigResult)
+	err := msg.UnmarshalResult(cfg)
+	if err != nil {
+		dcr.setWsFail(fmt.Errorf("error unmarshaling config response: %v", err))
+	}
+	// If the server is not of sufficient version to support the data API,
+	// BinSizes will be nil and we won't create any candle caches.
+	var smallestBin uint64
+	for _, durStr := range cfg.BinSizes {
+		dur, err := time.ParseDuration(durStr)
+		if err != nil {
+			dcr.setWsFail(fmt.Errorf("unparseable bin size in dcrdex config response: %q: %v", durStr, err))
+			break
+		}
+
+		var key candlestickKey
+		for k, d := range candlestickDurations {
+			if d == dur {
+				key = k
+				break
+			}
+		}
+		if key == "" {
+			log.Debugf("Skipping unknown candlestick duration %q", durStr)
```

Comment:
I guess we just cannot use the `5m` candlestick for dex.   Idea for having each market with a different bin set maybe?  We don't have an aggregate candlesticks view anyway, so seems like we can add `5m` and for explorer and have all the other exchanges missing this one.  This is fine for now though